feature: Add FairProjectConfig

ChristianTackeGSI opened the pull request:
Currently it has only one function: fair_get_git_version()

ChristianTackeGSI committed:
feature: Add FairProjectConfig

Currently it has only one function: fair_get_git_version()
Most of this function is from FairMQ's get_git_version().

ChristianTackeGSI committed:
ci: Include CMake 3.21 in Testing

CMake 3.21 has been released a few days ago.
Let's start testing with it on Linux.

ChristianTackeGSI commented:
Should we add a `--always` to the`git describe`?

dennisklein reviewed (commented):
https://github.com/FairRootGroup/FairCMakeModules/runs/3138860144?check_suite_focus=true#step:5:26 Is this behavior intended?

dennisklein commented:
> Should we add a `--always` to the`git describe`?

To be honest, I do not really understand what it does.

ChristianTackeGSI commented:
> https://github.com/FairRootGroup/FairCMakeModules/runs/3138860144?check_suite_focus=true#step:5:26 Is this behavior intended?

I guess, the problem is, that the github CI does not present a full git repository to the build process.

So we could either say: 
→ Okay, that's the case, where we want "don't touch variables" behaviour.

If that's the idea, we might consider to add `ERROR_VARIABLE` to `execute_process()` to hide the error messages. (I am unsure, if this is a good idea.)

Or:

> > Should we add a `--always` to the`git describe`?
> 
> To be honest, I do not really understand what it does.

If `git describe` can't find a good tag, it will return a shortened commit sha1 with that flag.

I have added a commit to show the effect:

```console
-- FairCMakeModules Version 0.2.0 (6c08dde)
```

For the github CI, we get some strange commit sha1. (I guess, it's some merge commit or something like that).

I am also unsure, whether this is really helpful. Maybe it's just "better than nothing"?

dennisklein commented:
> I am also unsure, whether this is really helpful. Maybe it's just "better than nothing"?

What would the code look like, if a project wanted to implement the behaviour "Get the git version, but fall back to something else, if a git version is not available"? I believe we should make sure that such a code snippet is not going to get overly complicated / un-readable because of the interface we choose for `fair_get_git_version()`.

dennisklein commented:
> > I am also unsure, whether this is really helpful. Maybe it's just "better than nothing"?
> 
> What would the code look like, if a project wanted to implement the behaviour "Get the git version, but fall back to something else, if a git version is not available"? I believe we should make sure that such a code snippet is not going to get overly complicated / un-readable because of the interface we choose for `fair_get_git_version()`.

```cmake
fair_get_git_version()
if(NOT PROJECT_GIT_VERSION)
  # get it from somewhere else
endif()
```

If there is no such logic in the project itself, and a package manager cannot present a git repo, it should also be straightforward to *inject* the version by modifying the source tree, e.g. with a package-level patch:

```diff
- fair_get_git_version()
+ set(PROJECT_GIT_VERSION "v42")
```

Looks simple enough, so from my side I would approve this.

> Maybe it's just "better than nothing"?

I would conclude, yes.

ChristianTackeGSI commented:
If we don't implement `--always`:
```cmake
set(PROJECT_GIT_VERSION "your fallback default here")
fair_get_git_version()

# Or
fair_get_git_version(OUTVAR_PREFIX yourprefix)
if(NOT DEFINED yourprefix_GIT_VERSIONyourprefix_GIT_VERSION)
   set(yourprefix_GIT_VERSION "your fallback")
   # Do something else
endif()
```

For the `--always` case the above code would only work for a non-existant / broekn `.git`. Otherwise it would always use the sha1.

In that case, one would do something like
```cmake
fair_get_git_version()
if(NOT DEFINED PROJECT_GIT_VERSION OR PROJECT_GIT_VERSION regex-match-operator "^[0-9a-z][0-9a-z][0-9a-z][0-9a-z]*$")
  # ...
endif()
```

Which obviously is ugly.

dennisklein commented:
> If that's the idea, we might consider to add ERROR_VARIABLE to execute_process() to hide the error messages. (I am unsure, if this is a good idea.)

Actually, what about having a flag `FATAL_ERROR` (or `REQUIRED`) which is by default `OFF` - meaning it would silence the error message and otherwise turn the output of `git describe` into a fatal cmake error on non-zero exit codes (possibly with some additional comments).

ChristianTackeGSI committed:
feature(FairProjectConfig): Add more options

Improve `fair_get_git_version()`:

1. Add `TAGS_ONLY` option

If git describe can't create a nice tag like answer, it used to
return an error and an empty string.

With `--always` it returns a (shortened) commit sha1.

Giving the TAGS_ONLY option will return to the previous
behaviour.

2. Add `REQUIRED` option

If `fair_get_git_version()` can't detect a suitable result,
it used to not touch any output variables (so that
developers could easily establish defaults) and possibly
just echo errors.  The later could be confusing to users.

`REQUIRED` will make this a fatal error, showing some
messages and the error output.

Not giving `REQUIRED` will hide all errors.

dennisklein commented on src/modules/FairProjectConfig.cmake:
```diff
@@ -0,0 +1,103 @@
+    set(git_always "--always")
+  endif()
+
+  if(GIT_FOUND AND EXISTS ${CMAKE_SOURCE_DIR}/.git)
```
```suggestion
  if(Git_FOUND AND EXISTS ${CMAKE_SOURCE_DIR}/.git)
```

The find module also sets `GIT_FOUND` but `Git_FOUND` is the more modern convention.

dennisklein replied:
the existance check for `.git` is fragile, consider using `git rev-parse --is-inside-work-tree` or similar (not sure how recent the `--is-inside-work-tree` flag is, also see e.g. https://remarkablemark.org/blog/2020/06/05/check-git-repository/ for simpler variant)

ChristianTackeGSI replied:
What about dropping the existence check completely and letting `git describe` handle it?

```console
CMake Error at src/modules/FairProjectConfig.cmake:91 (message):
  git describe failed (return value 128):

  fatal: not a git repository (or any parent up to mount point /)

  Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```

dennisklein commented on src/modules/FairProjectConfig.cmake:
```diff
@@ -0,0 +1,103 @@
+  endif()
+
+  if(GIT_FOUND AND EXISTS ${CMAKE_SOURCE_DIR}/.git)
+    execute_process(COMMAND ${GIT_EXECUTABLE} describe --tags --dirty ${git_always} --match "v*"
```
```suggestion
    # TODO use this form once this module requires CMake 3.14
    # execute_process(COMMAND $<TARGET_FILE:Git::Git> describe --tags --dirty ${git_always} --match "v*"
    execute_process(COMMAND ${GIT_EXECUTABLE} describe --tags --dirty ${git_always} --match "v*"
```

ChristianTackeGSI committed:
fix(FairProjectConfig): Handle Edge Cases Better

* Use Git_FOUND instead of GIT_FOUND
* Prepare for CMake 3.14
* Let git describe handle non existent git repos (and do
  not depend on `.git` implementation detail).

Co-authored-by: Dennis Klein <d.klein@gsi.de>
